bellatrix.playwright/src/main/java/solutions/bellatrix/playwright/infrastructure/BrowserLifecyclePlugin.java

```diff
@@ -0,0 +1,196 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package solutions.bellatrix.playwright.infrastructure;
+
+import solutions.bellatrix.core.plugins.Plugin;
+import solutions.bellatrix.core.plugins.TestResult;
+import solutions.bellatrix.core.utilities.DebugInformation;
+import solutions.bellatrix.playwright.utilities.Settings;
+
+import java.lang.reflect.Method;
+import java.util.Objects;
+
+public class BrowserLifecyclePlugin extends Plugin {
+    private static final ThreadLocal<BrowserConfiguration> CURRENT_BROWSER_CONFIGURATION;
+    private static final ThreadLocal<BrowserConfiguration> PREVIOUS_BROWSER_CONFIGURATION;
+    private static final ThreadLocal<Boolean> IS_BROWSER_STARTED_DURING_PRE_BEFORE_CLASS;
+    private static final ThreadLocal<Boolean> IS_BROWSER_STARTED_CORRECTLY;
+
+    static {
+        CURRENT_BROWSER_CONFIGURATION = new ThreadLocal<>();
+        PREVIOUS_BROWSER_CONFIGURATION = new ThreadLocal<>();
+        IS_BROWSER_STARTED_DURING_PRE_BEFORE_CLASS = new ThreadLocal<>();
+        IS_BROWSER_STARTED_CORRECTLY = new ThreadLocal<>();
+    }
+
+    @Override
+    public void preBeforeClass(Class type) {
+        if (Settings.web().getExecutionType().equals("regular")) {
+            currentConfiguration(getExecutionBrowserClassLevel(type));
+            if (shouldRestartBrowser()) {
+                shutdownBrowser();
+                startBrowser();
+                // TODO: maybe we can simplify and remove this parameter.
+                isBrowserStartedDuringPreBeforeClass(true);
+            } else {
+                isBrowserStartedDuringPreBeforeClass(false);
+            }
+        } else {
+            isBrowserStartedDuringPreBeforeClass(false);
+        }
+
+        super.preBeforeClass(type);
+    }
+
+    @Override
+    public void postAfterClass(Class type) {
+        shutdownBrowser();
+        isBrowserStartedDuringPreBeforeClass(false);
+        super.preAfterClass(type);
+    }
+
+    @Override
+    public void preBeforeTest(TestResult testResult, Method memberInfo) {
+        currentConfiguration(getBrowserConfiguration(memberInfo));
+
+        if (!isBrowserStartedDuringPreBeforeClass()) {
+            if (shouldRestartBrowser()) {
+                startBrowser();
+            }
+        }
+
+        isBrowserStartedDuringPreBeforeClass(false);
+    }
+
+    @Override
+    public void beforeTestFailed(Exception ex) throws Exception {
+        throw ex;
+    }
+
+    @Override
+    public void postAfterTest(TestResult testResult, Method memberInfo, Throwable failedTestException) {
+
+        if (currentConfiguration().lifecycle() == Lifecycle.REUSE_IF_STARTED) {
+            PlaywrightService.restartBrowserContext();
```

Comment:
isn't this `PlaywrightService` the only difference with the original BrwoserLifecyclePlugin? Can't we somehow abstract the service and reuse the code instead of copy-paste?